Add tests for `GetInterfaceMap` on static interface methods

hamarb123 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/90351.
Includes the currently single failing case from the comment https://github.com/dotnet/runtime/issues/90351#issuecomment-1676179889 - therefore I propose that all (/ or at least most) of the combinations are useful since they actually found a bug.
This probably can't be merged until it's fixed, or maybe that just needs to be disabled for now, but that's alright lol.

The failing case is the case which checks `C2Implicit<string>`'s implementations of `I1<string>.G<T>`. Previously, until recently when some other unrelated (to this) changes to static interface methods were made, many more cases were broken.

Related failure issues, which have had their tests disabled:
- https://github.com/dotnet/runtime/issues/90863
- https://github.com/dotnet/runtime/issues/91027

cc @jkotas

hamarb123 committed:
Add unit tests for #90351

hamarb123 committed:
Merge remote-tracking branch 'upstream' into main5

jkotas commented:
>maybe that just needs to be disabled for now, but that's alright lol.

I would recommend disabling the test against a new issue.

hamarb123 commented:
> I would recommend disabling the test against a new issue.

Is there a proper way to do this, or do I just need to skip the test manually? I ask because it's using an interesting setup, where it returns an `IEnumerable<object[]>` using `yield` as data for a `[Theory]`; presumably there's no way to do an `[ActiveIssue]` for just 1 of those cases?

jkotas commented:
Comment out the failing case and add `// [ActiveIssue("https://github.com/dotnet/runtime/issues/12345")]` in front of it.

hamarb123 committed:
Add [ActiveIssue] to skip over failing test case

hamarb123 commented:
I also need to exclude some tests on mono by the looks of it, I will add an `[ActiveIssue]` for that when I make an issue for that too.

hamarb123 committed:
Add some more DIM test cases

• Would be good to test these anyway
• Added to see if the failures on mono are with static interface methods only, or to do with default implementation methods

hamarb123 committed:
Fix added DIM tests

hamarb123 committed:
Disable some of the tests for mono

• Haven't opened an issue for this yet - will follow up with an issue & commit to add the issue link

hamarb123 commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.DUPLICATE_EXTENTS_DATA.cs:
```diff
@@ -0,0 +1,19 @@
+            internal long ByteCount;
+        }
+    }
+}
```
what, why is this here lol

hamarb123 committed:
Merge remote-tracking branch 'upstream/main' into main5

jkotas commented on src/libraries/System.Runtime/tests/System/Type/TypeTests.cs:
```diff
@@ -1070,6 +1097,148 @@ public static IEnumerable<object[]> GetInterfaceMap_TestData()
+                }
+
+                // Check we have the expected implementation for the level 3 interfaces (abstract explicit implementations - I1.M and I1.G methods)
+                bool isMono = typeof(object).Assembly.GetType("Mono.RuntimeStructs") != null; // [ActiveIssue("https://github.com/dotnet/runtime/issues/TODO")]
+                if (!isMono && classType.Index >= 3)
```
```suggestion
[ActiveIssue("https://github.com/dotnet/runtime/issues/TODO")]
                if (!PlatformDetection.IsMonoRuntime && classType.Index >= 3)
```

trylek reviewed (approved):
Looks great to me, thank you!

jkotas committed:
Update src/libraries/System.Runtime/tests/System/Type/TypeTests.cs

jkotas commented:
If there is nothing else to do, please flip the PR from Draft to Ready to review.

hamarb123 committed:
Update TypeTests.cs

hamarb123 commented:
> If there is nothing else to do, please flip the PR from Draft to Ready to review.

I haven't opened the other issue yet, if you're alright with merging it without that issue then feel free, or if you wanted to open it that would be great since I'm struggling to run mono locally and therefore write a useful reproduction.

jkotas commented:
For the tracking issue, it is fine to just link to the test that is disabled. You do not need to go the extra mile to extra the test into a small repro.

hamarb123 committed:
Add test number

jkotas commented:
/azp run runtime-extra-platforms

jkotas commented:
Native aot compiler fails on the new tests with:
```
Process terminated. Assertion failed.
   at Internal.TypeSystem.MetadataVirtualMethodAlgorithm.ResolveInterfaceMethodToVirtualMethodOnType(MethodDesc interfaceMethod, MetadataType currentType) in /_/src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs:line 615
   at ILCompiler.LazyGenericsSupport.GraphBuilder.LookForVirtualOverrides(EcmaMethod method) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/GraphBuilder.cs:line 129
   at ILCompiler.LazyGenericsSupport.GraphBuilder..ctor(EcmaModule assembly) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/GraphBuilder.cs:line 62
   at ILCompiler.LazyGenericsSupport.CycleInfoHashtable.CreateValueFromKey(EcmaModule key) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 52
   at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
   at ILCompiler.LazyGenericsSupport.GenericCycleDetector.FormsCycle(TypeSystemEntity entity, ModuleCycleInfo& cycleInfo) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 194
   at ILCompiler.LazyGenericsSupport.GenericCycleDetector.DetectCycle(TypeSystemEntity owner, TypeSystemEntity referent) in /_/src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs:line 230
   at Internal.IL.ILImporter.GetMethodEntrypoint(MethodDesc method) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 257
   at Internal.IL.ILImporter.ImportCall(ILOpcode opcode, Int32 token) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 642
   at Internal.IL.ILImporter.ImportBasicBlock(BasicBlock basicBlock) in /_/src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs:line 424
   at Internal.IL.ILImporter.ImportBasicBlocks() in /_/src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs:line 308
   at Internal.IL.ILImporter.Import() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs:line 166
   at ILCompiler.ILScanner.CompileSingleMethod(ScannedMethodNode methodCodeNodeNeedingCode) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 122
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`2.<ForWorker>b__1(RangeWorker& currentWorker, Int64 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
```
I am going to push a fix for this crash.

jkotas committed:
Fix native AOT compiler crash

jkotas commented:
/azp run runtime-extra-platforms

jkotas reviewed (approved):
Thank you!
